src/core/exec-invoke.c

```diff
@@ -4382,6 +4377,13 @@ int exec_invoke(
                 return log_exec_error_errno(context, params, r, "Failed to set up standard error output: %m");
         }
 
+        /* Now that stdin/stdout are opened, properly initialize it with our desired settings. (Note: this is
+         * a "constructive" reset, it prepares things for us to use. This is different from the "destructive"
+         * TTY reset further up.) */
+        if (context->tty_reset)
+                (void) terminal_reset_defensive(STDOUT_FILENO, /* switch_to_text= */ false);
```

Comment:
Hmm, I don't think moving this makes too much sense. We need to know whether/which of stdin/stdout/stderr is connected to the tty, but here we just heedlessly operate on stdout. `setup_output` even contains a comment like this:

> We don't reset the terminal if this is just about output

Reply:
Well, the thing though is that part of the reset is writing ansi sequences to an fd. And since only stdout can be assumed to be open for write we must use stdout to write the ansi sequence.

hence, terminal reset really is a concept tied to stdout, not so much to stdin.

And the other thing is that we nowadays support passing in pre-opened fds for stdin/stdout, and we really should reset the tty if we are told so if those fds refer to a tty. hence I think we should move this out of the code that opens ttys, and into code that handles bother ttys we open and those we get handed in.

i'll extend the comments/commit msgs about this.

Reply:
But people are free to specify StandardInput=tty yet StandardOutput=file/null/..., no? This would break miserably in such cases?

Reply:
Still, I think we should conditionalize these calls based on context->std_output/ExecOutput. It's really more obvious that way.

Also, should we reset stderr if it's not the same as stdout?

Reply:
but condition on what. if somebody gives us stdin/stdout as fds, i think we should run this stuff.

Reply:
I am not sure what precisely you are asking for now. Moving the initialization back doesn't really make sense, because we shouldn't write ansi sequences to standard input...

Reply:
I mean, I still think we should confine the calls in `setup_output` with `ExecOutput` taken into consideration. And we should reset stderr too if it's connected to a different tty.

`terminal_reset_ioctl` might log about ENOTTY BTW, which can totally be avoided (even though it's only debug level)